Create initial Haskell basics/operators page

TrueBoxGuy opened the pull request:
Closes #25. 

* Emphasise that operators are infix binary functions
* Explain how to use ordinary function identifiers as operators
* Explain usage of operators in prefix, including partial application
* Define and show examples of finding and using operator fixity
* Show a variety of custom operator and fixity declarations

TrueBoxGuy committed:
Create initial Haskell basics/operators page

* Emphasise that operators are infix binary functions
* Explain how to use ordinary function identifiers as operators
* Explain usage of operators in prefix, including partial application
* Define and show examples of finding and using operator fixity
* Show a variety of custom operator and fixity declarations

TrueBoxGuy commented on en/lessons/basics/operators.md:
````diff
@@ -0,0 +1,164 @@
+regular functions used in infix mode.
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> :info (+)
+infixl 6 +
````
Perhaps put something like `[...]` here to let the reader know text was skipped.

TrueBoxGuy replied:
The text above somewhat indicates that output was skipped, but I'd like to know whether others think this is clear enough.

ulysses4ever replied:
Adding `...` seems like a good idea to me, although not critical.

TrueBoxGuy commented on en/lessons/basics/operators.md:
````diff
@@ -0,0 +1,164 @@
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> (^ (3 - 1)) . subtract 3818 $ 4
+14546596
+ghci> (1729 + 1) * 121 `div` 2
````
Ensure that numbers are used in a way that makes the order of operations easy to deduce. For example here, the fact that the output is odd indicates that divides is of lower precedence than multiplication.

TrueBoxGuy commented on en/lessons/basics/operators.md:
````diff
@@ -0,0 +1,164 @@
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> (+) 2 3
+5
+ghci> (5 +) . (* 2) $ 8
````
Pick operators for which the side on which the expression is would matter.

TrueBoxGuy commented on en/lessons/basics/operators.md:
````diff
@@ -0,0 +1,164 @@
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> 2010 `mod` 1998
+12
+ghci> (`div` 3) . (* 1471684) $ 5
````
The usage of the concept of interchange may be confusing for the reader, given that this process is applied twice here: the example makes an ordinary function identifier into an operator, and then back into a prefix function that is partially applied.

TrueBoxGuy commented:
I'm not sure whether I've focussed enough on terminology, or whether it would get in the way of the lesson.

TrueBoxGuy commented:
![image](https://user-images.githubusercontent.com/17554188/126909561-6c3b2275-3850-4fe4-9496-f14b9cc37481.png)
Here's an image of the page.

soupi commented:
This is a pretty quality explanation.
I have a few comments:

- In a few places, the word "brackets" (`[]`) is used instead of "parenthesis" (`()`)
- Consider explicitly mentioning how partially applied operators are translated so the positions are more obvious. e.g.
    - `(<=)` -> `\a b -> a <= b`
    - `(1 <=)` -> `\a -> 1 <= a`
    - `(<= 1)` -> `\a -> a <= 1`

- I would suggest adding a type signature when defining an operator in at least one instance just to make sure it is covered
- Consider using this as an opportunity to discuss `++` as an example of right associativity and why it's faster than left associativity (for that case), and why `++` is slow in general (though this could be out of scope here, up to you)
- I would suggest adding a note that recommends to avoid defining operators because they make code harder to read - names usually give clues regarding what the function does and operators don't do that usually, and maybe suggest to make operators aliases to normal functions
- Consider linking to the [operators glossary](https://www.fpcomplete.com/haskell/tutorial/operators/) if it make sense

Unrelated: I'm not a fan of using ligatures and unicode symbols in code. I think this can result in confusion for the reader and make the content less accessible. An example of this is `<=` being rendered as `≤` in this document.
I suggest removing ligatures and unicode from the content.  @Kleidukos

Kleidukos commented:
Agreed to disable ligatures

TrueBoxGuy commented:
@Kleidukos do you have any plans on how we should link to FP complete? Can I just put it at the beginning of the lesson?

Kleidukos commented:
@TrueBoxGuy Michael allowed us to display FPCo content on Haskell School under condition of referencing the original content on the FPCo website

TrueBoxGuy committed:
Respond to PR review

* Refer to FPComplete's Operator Glossary
* Modify numbers so order of operations is easier to deduce mentally
* Replace incorrect usage of word "brackets" with "parentheses"
* Display equivalent anonymous functions for parenthesised operators
* Explain how operator associativity affects output, program validity, and speed
* Warn reader about excessively defining operators
* Display example type signature for an operator

JonathanLorimer commented on en/lessons/basics/operators.md:
```diff
@@ -0,0 +1,185 @@
+
+## Using Operators
+
+Operators in Haskell are functions that can be placed between two arguments: they can
+return a value of any type and their names are either a string of symbols or ordinary
+function names surrounded in backticks. Using infix notation makes it easier to write
```
I don't know if "they can return a value of any type" is necessary here. I think the sentence would be more clear if it read something like this:
```suggestion
Operators in Haskell are functions that can be placed between two arguments. There are two cases where you can use a function as an operator: when the function name only consists of symbols or when you surround the function name in backticks. Using infix notation makes it easier to write expressions containing many common functions:
```

TrueBoxGuy replied:
I agree. The fact that any function can be an operator is made clear later.

JonathanLorimer commented on en/lessons/basics/operators.md:
````diff
@@ -0,0 +1,185 @@
+ghci> - (7 + 57843)
+-57850
+```
+__Note__: negate (`-`) is the only unary operator in Haskell.
````
this might be a bit confusing since the `-` is overloaded and can be used for subtraction too.

TrueBoxGuy replied:
Furthermore, my note doesn't say what use knowing that "-" is also a unary operator is: I may give an example of how it makes sectioning more difficult.

JonathanLorimer commented on en/lessons/basics/operators.md:
````diff
@@ -0,0 +1,185 @@
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> abs . subtract 119 $ 2 + 2
+115
+ghci> (`mod` 930233356) . negate $ 4
````
partial application of operators might prove a bit confusing for a first example. If anything I suggest breaking this out into a different codeblock and explain what you are trying to demonstrate here.

JonathanLorimer commented on en/lessons/basics/operators.md:
````diff
@@ -0,0 +1,185 @@
+
+### Interchanging Operators and Functions
+
+To use an operator symbol as if it were a regular function, surround it in parentheses.
+Furthermore, you can create partially applied operators, called sections, by putting
+an expression on only one side of the operator:
+
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> (+) 2 3
+5
+ghci> (5 <=) . (* 2) $ 3
+True
+```
````
Ah yeah, this was the kind of explanation I was looking for in the example above.

TrueBoxGuy replied:
Is it fine in this order?

JonathanLorimer replied:
It might be nice to briefly mention this in the intro to give context, but I leave that decision up to your discretion.

JonathanLorimer commented on en/lessons/basics/operators.md:
````diff
@@ -0,0 +1,185 @@
+True
+```
+
+The parenthesised operators in the above example can be treated as equivalent to the following
+anonymous functions:
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> (\a b -> a + b) 2 3
+5
+ghci> (\a -> 5 <= a) . (\a -> a * 2) $ 8
+True
+```
````
nice explanation

TrueBoxGuy replied:
You can thank soupi for it :).

JonathanLorimer commented on en/lessons/basics/operators.md:
````diff
@@ -0,0 +1,185 @@
+Conversely, to make a regular function identifier an operator, surround it in backticks:
+
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> 2010 `mod` 1998
+12
````
Maybe show `mod` used prefix first. I find the contrast can be useful for learning.

JonathanLorimer commented on en/lessons/basics/operators.md:
```diff
@@ -0,0 +1,185 @@
+
+### Operator Precedence and Associativity
+
+An operator's precedence and associativity combined is called its fixity, and it
+controls how the operator is parsed in an expression. To find out the fixity of an
+operator, look through `ghci`'s `:info` or the documentation for the operator:
+if fixity information isn't displayed, it defaults to `infixl 9`. The same applies to
+regular functions used in infix mode.
```
Can you explain what you mean by associativity? i.e. that the prefix of `r` or `l` in the `infixl 9` declaration indicates the direction in which it associates. A really good illustration of this is the `esqueleto` experimental syntax operator `:&`  which allows you to pattern match and ignore all arguments to the left because of its associativity:

```haskell
(a :& b :& c :& d :& e) == (_ :& d :& e)
```

TrueBoxGuy replied:
I cover a process that explains how associativity works and then provide a couple examples later on. Do you think these examples build enough intuition? 

I feel like the most intuitive examples of associativity are in data constructors, such as that for list, but I'm not sure I should mention that here.

JonathanLorimer replied:
Yeah, this subject does get more advanced very quickly if you delve into to much detail. Again, I leave it to your discretion.

JonathanLorimer commented on en/lessons/basics/operators.md:
```diff
@@ -0,0 +1,185 @@
+<interactive>:1:1: error:
+    Precedence parsing error
+        cannot mix ‘==’ [infix 4] and ‘==’ [infix 4] in the same infix expression
+ghci> 0 <> 1 + (-1)
+<interactive>:2:1: error:
+    Precedence parsing error
+        cannot mix ‘<>’ [infixr 6] and ‘+’ [infixl 6] in the same infix expression
```
nice example

TrueBoxGuy replied:
Thanks. It's a small easter egg because it can typecheck if you bracket it: `(0 <> 1) + (-1) :: Sum Integer`.

TrueBoxGuy committed:
Respond to PR review

* Rewrite definition of an operator
* Mention difficulty in sectioning the subtraction operator
* Provide adjacent examples of using prefix and infix notation

TrueBoxGuy committed:
Clarify that example is meant to fascinate

Clarify that reader isn't supposed to understand usage of operator sections and multiple operators in first codeblock

JonathanLorimer reviewed (approved):
Looks good to me

TrueBoxGuy committed:
Warn readers about overusing fixity

Kleidukos reviewed (approved):
LGTM
